### PR TITLE
nixos/openssh: Add sntrup761x25519-sha512 kexAlgo

### PR DESCRIPTION
**Pull request by mweinelt**
###### Description of changes

Introduced in OpenSSH 9.0 it became the part of the default kexAlgorithm
selection, visibile in sshd_config(5).

It is also enabled by default in the OpenSSH client, as can be seen from

$ ssh -Q KexAlgorithms

Also clarifies that we use the referenced documents as the lower bound,
given that they haven't been updated for 5-7y.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/ssh/sshd.nix [hunk 1]
```diff
@@ -301,7 +302,7 @@ in
           Allowed key exchange algorithms
           </para>
           <para>
-          Defaults to recommended settings from both
+          Uses the lower bound recommended in both
           <link xlink:href="https://stribika.github.io/2015/01/04/secure-secure-shell.html" />
           and
           <link xlink:href="https://infosec.mozilla.org/guidelines/openssh#modern-openssh-67" />
```
Comment (infinisil) [inline nixos/modules/services/networking/ssh/sshd.nix:308]: I don't see `sntrup761x25519-sha512@openssh.com` suggested in either of those links.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/ssh/sshd.nix (hunk 1 above)
Comment (mweinelt) [reply nixos/modules/services/networking/ssh/sshd.nix:308]: Yeah, because neither of them has been updated in years.

Comment (infinisil) [reply nixos/modules/services/networking/ssh/sshd.nix:308]: I see, that makes sense. However I'm not sure if we should add this as a default algorithm yet. I can barely find any information on it. Do you know more about this algorithm/its history/why it should be safe to enable by default?

Comment (zeri42) [reply nixos/modules/services/networking/ssh/sshd.nix:308]: The [OpenSSH Release Notes](https://www.openssh.com/releasenotes.html) seem clear to me.
 
```
* [ssh(1)](https://man.openbsd.org/ssh.1), [sshd(8)](https://man.openbsd.org/sshd.8): use the hybrid Streamlined NTRU Prime + x25519 key
   exchange method by default ("sntrup761x25519-sha512@openssh.com").
   The NTRU algorithm is believed to resist attacks enabled by future
   quantum computers and is paired with the X25519 ECDH key exchange
   (the previous default) as a backstop against any weaknesses in
   NTRU Prime that may be discovered in the future. The combination
   ensures that the hybrid exchange offers at least as good security
   as the status quo.
```

The idea here is to do a key exchange using both algorithms (ntruprime and x25519) in a way that the key exchange remains secure unless both schemes are compromised. The exact method seems to be to do a [sntru761 elliptic curve key exchange and a key exchange using x25519 and concatenate the results before feeding them to a hash function:
](https://github.com/Sineware/openssh/blob/279261e1ea8150c7c64ab5fe7cb4a4ea17acbb29/kexsntrup4591761x25519.c#L119-L141) (I didn't want to install CVS for the original source feel free to flame me if the original has a different sha256)


 NTRU Prime (not to be confused with NTRU) is one of nine candidates (counting only key encapsulation mechanisms and ignoring signature schemes) for [NISTs current effort to create the next standard](https://csrc.nist.gov/Projects/post-quantum-cryptography/round-3-submissions). This should be fine performance wise too, as structured lattice schemes (Kyber + NTRU + NTRU Prime ... bet I forgot some half of them are lattice) have really good performance.  [Ths benchmark for all nist-pqc candidates on a Cortex M4](https://github.com/mupq/pqm4/blob/master/benchmarks.md ) is the only source I could think of the top of my head. Note: NISTs wanted KEMs (key encapsulation mechanisms) submitted that's why they are called that, but they behave like a key exchange that you can also use for public key encryption. Long story: short the algorithm is used in a way that can only improve security and it is very fast compared to other pqc schemes. This is why it doesn't hurt too much to add an additional key exchange. The picture would be different if the scheme was very slow (SIKE) or the public key would be like a million+ bytes (classic McElice).

Comment (infinisil) [reply nixos/modules/services/networking/ssh/sshd.nix:308]: Sounds good enough to me, thanks for the detailed info, I'll merge :)

**Review by infinisil — APPROVED** [review]

**Pull request merged**